spark/src/main/java/org/apache/iceberg/actions/ExpireSnapshotsAction.java

```diff
@@ -146,32 +147,48 @@ public ExpireSnapshotsAction deleteWith(Consumer<String> newDeleteFunc) {
     return this;
   }
 
-  @Override
-  public ExpireSnapshotsActionResult execute() {
-    // Metadata before Expiration
-    Dataset<Row>  originalFiles = buildValidFileDF(ops.current());
+  /**
+   * Expires snapshots and commits the changes to the table, returning a Dataset of files to delete.
+   * <p>
+   * This does not delete data files. To delete data files, run {@link #execute()}.
+   * <p>
+   * This may be called before or after {@link #execute()} is called to return the expired file list.
+   *
+   * @return a Dataset of files that are no longer referenced by the table
+   */
+  public Dataset<Row> expire() {
+    if (expiredFiles == null) {
```

Comment:
Should we be synchronizing on expired files? Probably a really unlikely situation to be calling this object in a multi-threaded way or to call expire and execute concurrently ...

Reply:
I'd rather not over-complicate it. If there is a future use case for multiple threads we can do it then, but I really doubt it will be needed.